File: nxflix-agents-python/src/nxflix_agents/services/kanji.py
"""Kanji service for managing JLPT N1 kanji items."""

from nxflix_agents.models import KanjiItem


class KanjiService:
    """Service for managing kanji items in memory."""

    def __init__(self) -> None:
        self._kanji_items: dict[str, KanjiItem] = {}
        self._by_character: dict[str, str] = {}
        self._by_stroke_count: dict[int, list[str]] = {}

    def add_kanji_item(self, kanji: KanjiItem) -> None:
        """Add a kanji item to the service."""
        self._kanji_items[kanji.id] = kanji
        self._by_character[kanji.character] = kanji.id

        if kanji.stroke_count not in self._by_stroke_count:
            self._by_stroke_count[kanji.stroke_count] = []
        self._by_stroke_count[kanji.stroke_count].append(kanji.id)

    def get_kanji_item(self, kanji_id: str) -> KanjiItem | None:
        """Get a kanji item by ID."""
        return self._kanji_items.get(kanji_id)

    def get_kanji_by_character(self, character: str) -> KanjiItem | None:
        """Get a kanji item by its character."""
        kanji_id = self._by_character.get(character)
        return self._kanji_items.get(kanji_id) if kanji_id else None

    def get_all_kanji_items(self) -> list[KanjiItem]:
        """Get all kanji items."""
        return list(self._kanji_items.values())

    def get_kanji_by_stroke_count(self, stroke_count: int) -> list[KanjiItem]:
        """Get kanji items by stroke count."""
        ids = self._by_stroke_count.get(stroke_count, [])
        return [self._kanji_items[id_] for id_ in ids if id_ in self._kanji_items]

    def get_kanji_items_by_ids(self, ids: list[str]) -> list[KanjiItem]:
        """Get multiple kanji items by their IDs."""
        return [
            self._kanji_items[id_] for id_ in ids if id_ in self._kanji_items
        ]

    def search_kanji(self, query: str) -> list[KanjiItem]:
        """Search kanji by character, meaning, or readings."""
        lower_query = query.lower()
        results: list[KanjiItem] = []

        for kanji in self._kanji_items.values():
            if (
                query in kanji.character
                or any(m.lower().find(lower_query) != -1 for m in kanji.meanings)
                or any(query in r for r in kanji.onyomi)
                or any(query in r for r in kanji.kunyomi)
            ):
                results.append(kanji)

        return results

    @property
    def count(self) -> int:
        """Get the number of kanji items."""
        return len(self._kanji_items)

    @property
    def stroke_counts(self) -> list[int]:
        """Get all available stroke counts."""
        return sorted(self._by_stroke_count.keys())

File: nxflix-agents-ts/src/services/kanji.ts
import type { KanjiItem } from '../models/kanji.js';

export class KanjiService {
  private kanjiItems: Map<string, KanjiItem> = new Map();
  private byCharacter: Map<string, string> = new Map();
  private byStrokeCount: Map<number, string[]> = new Map();

  addKanjiItem(kanji: KanjiItem): void {
    this.kanjiItems.set(kanji.id, kanji);
    this.byCharacter.set(kanji.character, kanji.id);

    const strokeList = this.byStrokeCount.get(kanji.strokeCount) ?? [];
    strokeList.push(kanji.id);
    this.byStrokeCount.set(kanji.strokeCount, strokeList);
  }

  getKanjiItem(kanjiId: string): KanjiItem | undefined {
    return this.kanjiItems.get(kanjiId);
  }

  getKanjiByCharacter(character: string): KanjiItem | undefined {
    const id = this.byCharacter.get(character);
    return id ? this.kanjiItems.get(id) : undefined;
  }

  getAllKanjiItems(): KanjiItem[] {
    return Array.from(this.kanjiItems.values());
  }

  getKanjiByStrokeCount(strokeCount: number): KanjiItem[] {
    const ids = this.byStrokeCount.get(strokeCount) ?? [];
    return ids
      .map((id) => this.kanjiItems.get(id))
      .filter((k): k is KanjiItem => k !== undefined);
  }

  getKanjiItemsByIds(ids: string[]): KanjiItem[] {
    return ids
      .map((id) => this.kanjiItems.get(id))
      .filter((k): k is KanjiItem => k !== undefined);
  }

  searchKanji(query: string): KanjiItem[] {
    const lowerQuery = query.toLowerCase();
    const results: KanjiItem[] = [];

    for (const kanji of this.kanjiItems.values()) {
      if (
        kanji.character.includes(query) ||
        kanji.meanings.some((m) => m.toLowerCase().includes(lowerQuery)) ||
        kanji.onyomi.some((r) => r.includes(query)) ||
        kanji.kunyomi.some((r) => r.includes(query))
      ) {
        results.push(kanji);
      }
    }

    return results;
  }

  get count(): number {
    return this.kanjiItems.size;
  }

  get strokeCounts(): number[] {
    return Array.from(this.byStrokeCount.keys()).sort((a, b) => a - b);
  }
}
